include/socket_offload_ops.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2017 Texas Instruments
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Socket Offload Redirect API
+ */
+
+#ifndef __SOCKET_OFFLOAD_OPS_H__
+#define __SOCKET_OFFLOAD_OPS_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <net/net_ip.h>
+
+/*
+ * It is assumed that these offload functions (except for init()) follow the
+ * POSIX socket API standard for arguments, return values and setting of errno.
+ */
+struct socket_offload {
+	/* Perform one time initialization for the offload device: */
+	void (*init)();
+	/* POSIX Socket Functions: */
+	int (*socket)(int family, int type, int proto);
+	int (*close)(int sock);
+	int (*accept)(int sock, struct sockaddr *addr, socklen_t *addrlen);
+	int (*bind)(int sock, const struct sockaddr *addr, socklen_t addrlen);
+	int (*listen)(int sock, int backlog);
+	int (*connect)(int sock, const struct sockaddr *addr,
+		       socklen_t addrlen);
+	int (*setsockopt)(int sock, int level, int optname,
+			  const void *optval, socklen_t optlen);
+	int (*getsockopt)(int sock, int level, int optname, void *optval,
+			  socklen_t *optlen);
+	ssize_t (*recv)(int sock, void *buf, size_t max_len, int flags);
```

Comment:
In POSIX, recv(...) == recfrom(..., NULL). I think we should rely on that property, and have only recvfrom/sendto virtual methods.

Reply:
Sounds logical.  I imagine a similar simplification can be done for send():  i.e., send() == sendto(..., NULL, 0).
Then, some simplification also for zsock_ and net_context_ APIs as well?